### PR TITLE
Replace black with ruff

### PR DESCRIPTION
**Pull request by sujuka99**

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -35,14 +35,19 @@ jobs:
...
+          if [[ "$RUNNER_OS" == "Linux" && "${{ matrix.python-version }}" == "3.10" ]]
+          then
+            echo "::add-matcher::.github/ruff-matcher.json"
+            poetry run ruff check . --config pyproject.toml --output-format text --no-fix
+          else
+            poetry run pre-commit run ruff --all-files --show-diff-on-failure
+          fi;
+
+      - name: Formatting (ruff)
+        run: poetry run pre-commit run ruff-format --all-files --show-diff-on-failure
```
Comment (raminqaf) [inline .github/workflows/ci.yaml:50]: Maybe you can also add some docs telling how people can configure ruff as their default formatter?

Comment (raminqaf) [reply .github/workflows/ci.yaml:50]: - https://plugins.jetbrains.com/plugin/20574-ruff
- https://github.com/astral-sh/ruff-vscode

Comment (raminqaf) [reply .github/workflows/ci.yaml:50]: This can be the first draft of contribution.md

Comment (sujuka99) [reply .github/workflows/ci.yaml:50]: Sure

File: docs/docs/developer/style.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# Style
+
+## Python
+
+To ensure a consistent Python code style, we use [Ruff](https://docs.astral.sh/ruff/) for both linting and formatting.
```
Comment (raminqaf) [inline docs/docs/developer/style.md:5]: Maybe pin the version here or link to the poetry file.

File: docs/docs/developer/style.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+## Markdown
+
+To ensure a consistent markdown style, we use [dprint](https://dprint.dev) to check and reformat.
+
+```shell
+dprint fmt
+```
+
+Use the [official documentation](https://dprint.dev/setup/) to set up dprint.
```
Comment (raminqaf) [inline docs/docs/developer/style.md:26]:
```suggestion
Use the [official documentation](https://dprint.dev/setup/) to setup dprint.
```

Comment (disrupted) [reply docs/docs/developer/style.md:26]: I believe set up is correct

File: docs/docs/developer/style.md
Comment (disrupted) [inline docs/docs/developer/style.md:1]: should we move this file to `CONTRIBUTING.md`?

File: docs/docs/developer/style.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+- [VSCode](https://marketplace.visualstudio.com/items?itemName=charliermarsh.ruff)
+- [PyCharm](https://plugins.jetbrains.com/plugin/20574-ruff)
+
+## Markdown
+
+To ensure a consistent markdown style, we use [dprint](https://dprint.dev) to check and reformat.
+
+```shell
+dprint fmt
```
Comment (disrupted) [inline docs/docs/developer/style.md:23]: I would just advice to stick to our git hooks, i.e. `pre-commit install`

Comment (sujuka99) [reply docs/docs/developer/style.md]: wdyt now?

File: docs/docs/developer/style.md
Comment (sujuka99) [reply docs/docs/developer/style.md:1]: I did create a `CONTRIBUTING.md` and a small script that copies it into the docs.

File: docs/docs/developer/style.md (hunk 2 above)
Comment (sujuka99) [reply docs/docs/developer/style.md]: I added a link to the poetry file a line or two below. Is it enough?

Manually keeping the version up to date wouldn't go too well imo and they don't have a version dropdown or tags like we do.

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -1,21 +1,22 @@
...
         pass_filenames: false
-      - id: black
-        name: black
-        entry: black
+      - id: ruff-format
+        name: ruff-format
+        entry: ruff format .
+        args: [ --config, pyproject.toml ]
         language: system
-        types_or: [python, pyi]
```
Comment (disrupted) [inline .pre-commit-config.yaml:16]: why did you change it and what's the difference to `types`?

File: .pre-commit-config.yaml [hunk 6]
```diff
@@ -1,21 +1,22 @@
...
         language: system
         types_or: [python]
         require_serial: true # run once for all files
         pass_filenames: false
-      - id: black
-        name: black
-        entry: black
+      - id: ruff-format
+        name: ruff-format
+        entry: ruff format .
```
Comment (disrupted) [inline .pre-commit-config.yaml:14]: I would prefer only formatting the files that are staged, otherwise it might fail if you have other local files

File: .pre-commit-config.yaml (hunk 5 above)
Comment (sujuka99) [reply .pre-commit-config.yaml:16]: Because we don't have `pyi` files. Also `pyright` already isn't looking at those, so I thought that we don't need it.

I could add `pyi` to both instead?

`types_or` works with an `OR`, so if any of the tags is present, the file is accepted.
`types` uses `AND`, so all tags need to be present in the same file

File: .pre-commit-config.yaml (hunk 6 above)
Comment (sujuka99) [reply .pre-commit-config.yaml:14]: Makes sense, but isn't that controlled from the CLI?

File: .pre-commit-config.yaml (hunk 5 above)
Comment (disrupted) [reply .pre-commit-config.yaml:16]: thanks for the explanation. I agree with removing `pyi`, we currently don't need it.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
+# Contributions
+
+**Welcome!** We are glad to have you visit our contributibg guide!
```
Comment (disrupted) [inline CONTRIBUTING.md:3]: typo in contribution

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+If you find any bugs or have suggestions for improvements, please open an [issue](https://github.com/bakdata/kpops/issues/new) and optionally a [pull request (PR)](https://github.com/bakdata/kpops/compare). In the case of a PR, we would appreciate it if you preface it with an issue outlining your goal and means of achieving it.
+
+## Style
+
+We advise that you stick to our `pre-commit` hooks, which you can run with `poetry run pre-commit run -a`. In order for `dprint` to work, you have to manually [install](#markdown) it locally. It will work in the CI, so it is also possible to manually carry out formatting changes flagged by `dprint` in the CI and skip installing it locally.
```
Comment (disrupted) [inline CONTRIBUTING.md:9]:
```suggestion
We advise that you stick to our `pre-commit` hooks for code linting, formatting, and auto-generation of our documentation and schemas. After you install them using `poetry run pre-commit install` they're triggered automatically during `git commit`. Additionally, you can manually invoke them with `poetry run pre-commit run -a`. In order for `dprint` to work, you have to manually [install](#markdown) it locally. It will work in the CI, so it is also possible to manually carry out formatting changes flagged by `dprint` in the CI and skip installing it locally.
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+Our configuration can be found in [KPOps](https://github.com/bakdata/kpops)' top-level `pyproject.toml`.
+
+**Editor integration**
+
+Below are listed existing Ruff plugins/extensions for some of the most popular python IDEs.
+
+- [VSCode](https://marketplace.visualstudio.com/items?itemName=charliermarsh.ruff)
+- [JetBrains](https://plugins.jetbrains.com/plugin/20574-ruff)
+
+If you cannot find your Editor of choices or you want something more custom, [`ruff-lsp`](https://github.com/astral-sh/ruff-lsp) enables Ruff to be used in any editor that supports the LSP
```
Comment (disrupted) [inline CONTRIBUTING.md:26]: I certainly appreciate the effort, however this can easily become outdated. Ideally, this should be covered by the official Ruff documentation. wdyt?

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+**Configuration**
+
+Our configuration can be found [here](https://github.com/bakdata/kpops/blob/main/dprint.json).
+
+**Editor integration**
+
+Below are listed existing `dprint` plugins/extensions for some of the most popular python IDEs.
+
+- [VSCode](https://marketplace.visualstudio.com/items?itemName=dprint.dprint)
+- [JetBrains](https://plugins.jetbrains.com/plugin/18192-dprint)
```
Comment (disrupted) [inline CONTRIBUTING.md:41]: similar, imo it's up to `dprint` to provide proper advice on editor integration.

File: hooks/gen_docs/gen_docs_contributing.py
Comment (disrupted) [inline hooks/gen_docs/gen_docs_contributing.py:1]: is it necessary to copy the file or can we use a symbolic link instead?

File: hooks/gen_docs/gen_docs_contributing.py [hunk 11]
```diff
@@ -0,0 +1,21 @@
+"""Copies ``CONTROBUTING.md`` from root to the docs."""
```
Comment (disrupted) [inline hooks/gen_docs/gen_docs_contributing.py:1]:
```suggestion
"""Copies ``CONTRIBUTING.md`` from root to the docs."""
```

**Review by disrupted — CHANGES_REQUESTED** [review]

File: docs/docs/developer/style.md (hunk 4 above)
Comment (disrupted) [reply docs/docs/developer/style.md]: looks good, commented small remarks :)

File: CONTRIBUTING.md (hunk 9 above)
Comment (sujuka99) [reply CONTRIBUTING.md:26]: You're right.

I could link [this](https://docs.astral.sh/ruff/editor-integrations/) page instead? Not sure if the change is in conflict with https://github.com/bakdata/kpops/pull/365#discussion_r1345401289 though. WDYT @raminqaf ?

File: CONTRIBUTING.md (hunk 10 above)
Comment (sujuka99) [reply CONTRIBUTING.md:41]: True, but I couldn't easily find such official resources. This one is very optional imo, so I'm fine with just leaving it out. I don't see much importance in having it outside of the `pre-commit` hook.

File: hooks/gen_docs/gen_docs_contributing.py
Comment (sujuka99) [reply hooks/gen_docs/gen_docs_contributing.py:1]: Didn't think of that, will try it out. I did try using snippets, but it didn't work.

File: CONTRIBUTING.md (hunk 9 above)
Comment (disrupted) [reply CONTRIBUTING.md:26]: link 404s for me

Comment (sujuka99) [reply CONTRIBUTING.md:26]: They changed it in `0.1.3` apparently :D

Here is the new one: https://docs.astral.sh/ruff/integrations/

File: docs/docs/stylesheets/extra.css [hunk 12]
```diff
@@ -13,6 +13,8 @@
 
     /* Color the links with the primary color */
     --md-typeset-a-color: var(--md-primary-fg-color);
+
+    --md-admonition-icon--wip: url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m13.78 15.3 6 6 2.11-2.16-6-6-2.11 2.16m3.72-5.2c-.39 0-.81-.05-1.14-.19L4.97 21.25l-2.11-2.11 7.41-7.4L8.5 9.96l-.72.7-1.45-1.41v2.86l-.7.7-3.52-3.56.7-.7h2.81l-1.4-1.41 3.56-3.56a2.976 2.976 0 0 1 4.22 0L9.89 5.74l1.41 1.4-.71.71 1.79 1.78 1.82-1.88c-.14-.33-.2-.75-.2-1.12a3.49 3.49 0 0 1 3.5-3.52c.59 0 1.11.14 1.58.42L16.41 6.2l1.5 1.5 2.67-2.67c.28.47.42.97.42 1.6 0 1.92-1.55 3.47-3.5 3.47Z"/></svg>')
```
Comment (raminqaf) [inline docs/docs/stylesheets/extra.css:17]: What is this?

File: CONTRIBUTING.md (hunk 9 above)
Comment (raminqaf) [reply CONTRIBUTING.md:26]: I am fine linking the this: [https://docs.astral.sh/ruff/integrations/](https://docs.astral.sh/ruff/integrations/)

File: hooks/gen_docs/gen_docs_contributing.py
Comment (raminqaf) [reply hooks/gen_docs/gen_docs_contributing.py:1]: If the script is the way to go it is cool to have another script that copy and paste the CHANGELOG.md for the docs :)

File: docs/docs/stylesheets/extra.css (hunk 12 above)
Comment (sujuka99) [reply docs/docs/stylesheets/extra.css:17]: An icon of a hammer and a wrench for the custom admonition. [tutorial](https://squidfunk.github.io/mkdocs-material/reference/admonitions/#custom-admonitions)

Comment (disrupted) [reply docs/docs/stylesheets/extra.css:17]: pied piper 🤣

File: hooks/gen_docs/gen_docs_contributing.py
Comment (sujuka99) [reply hooks/gen_docs/gen_docs_contributing.py:1]: I tried, read around and it doesn't seem possible.

Comment (sujuka99) [reply hooks/gen_docs/gen_docs_contributing.py:1]: @raminqaf where should the changelog live in the website?

Comment (disrupted) [reply hooks/gen_docs/gen_docs_contributing.py:1]: > I tried, read around and it doesn't seem possible.

that is symlinks don't work, right?

Comment (disrupted) [reply hooks/gen_docs/gen_docs_contributing.py:1]: the symlink could also be the other way around if it makes sense, i.e. the file lives in the docs and there's a symlink from the repository root to it

File: hooks/gen_docs/utils.py [hunk 13]
```diff
@@ -28,7 +28,7 @@ def copy_file(
             shutil.copyfileobj(file_source, file_dest)
             return
         if header is None:
-            header_encoded = f"<!-- This file is autogenerated by {Path(inspect.stack()[1].filename ).relative_to(ROOT)}. -->\n\n".encode()
+            header_encoded = f"<!-- This file is autogenerated by {Path(inspect.stack()[1].filename ).relative_to(ROOT)}. -->\n".encode()
```
Comment (disrupted) [inline hooks/gen_docs/utils.py:31]: what's with the extra space? does Ruff format it like this?
```suggestion
            header_encoded = f"<!-- This file is autogenerated by {Path(inspect.stack()[1].filename).relative_to(ROOT)}. -->\n".encode()
```

Comment (sujuka99) [reply hooks/gen_docs/utils.py:31]: Likely a missclick, it's not `ruff`

File: hooks/gen_docs/gen_docs_contributing.py
Comment (sujuka99) [reply hooks/gen_docs/gen_docs_contributing.py:1]: How could I pass those symlinks to github for the changelog and contributing?

Comment (disrupted) [reply hooks/gen_docs/gen_docs_contributing.py:1]: git supports tracking symlinks natively. I imagine something like this should work:
```sh
mv CONTRIBUTING.md docs/.../contributing.md
ln -s docs/.../contributing.md CONTRIBUTING.md
git add docs/.../contributing.md CONTRIBUTING.md
```

File: .github/workflows/release.yaml [hunk 14]
```diff
@@ -22,6 +22,7 @@ jobs:
       poetry-version: "1.5.1"
       changelog: true
       changelog-config: "./.github/changelog-config.json"
+      changelog-file: "./docs/docs/developer/contributing.md"
```
Comment (disrupted) [inline .github/workflows/release.yaml:25]: shouldn't it be changelog.md?

File: .pre-commit-config.yaml [hunk 15]
```diff
@@ -1,21 +1,20 @@
...
-        entry: black
+      - id: ruff-format
+        name: ruff-format
+        entry: ruff format
+        args: [ --force-exclude, --config, pyproject.toml ]
         language: system
-        types_or: [python, pyi]
+        types_or: [python]
         require_serial: true # run once for all files
-        exclude: ^tests/.*snapshots/
```
Comment (disrupted) [inline .pre-commit-config.yaml:18]: don't we need to continue excluding the snapshots? it's unnecessary to format them since they're generated

File: docs/docs/developer/auto-generation.md [hunk 16]
```diff
@@ -37,6 +37,10 @@ Generated
 - `pipeline-components/*\.yaml` -- All single-component pipeline definitions and one big (complete) `pipeline.yaml` that contains all of them.
 - `pipeline-defaults/*\.yaml` -- All single-component defaults definitions and one big (complete) `defaults.yaml` that contains all of them.
 
+#### [Contributing](../developer/contributing.md)
+
+Copied from [`CONTRIBUTING.md`](https://github.com/bakdata/kpops/blob/main/CONTRIBUTING.md)
+
```
Comment (disrupted) [inline docs/docs/developer/auto-generation.md:43]: still relevant?

File: docs/docs/developer/contributing.md [hunk 17]
```diff
@@ -0,0 +1,19 @@
+# Contribution
+
+**Welcome!** We are glad to have you visit our contributibg guide!
```
Comment (disrupted) [inline docs/docs/developer/contributing.md:3]: typo

File: docs/docs/developer/contributing.md [hunk 18]
```diff
@@ -0,0 +1,19 @@
+# Contribution
```
Comment (disrupted) [inline docs/docs/developer/contributing.md:1]: perhaps _How to contribute?_

File: docs/docs/developer/contributing.md
Comment (disrupted) [inline docs/docs/developer/contributing.md:1]: this is great!

File: docs/docs/stylesheets/extra.css [hunk 19]
```diff
@@ -22,3 +24,18 @@
...
+.md-typeset details.wip {
+border-color: rgb(160, 45, 0);
+}
+.md-typeset .wip > .admonition-title,
+.md-typeset .wip > summary {
+background-color: rgb(255, 195, 172);
+}
+.md-typeset .wip > .admonition-title::before,
+.md-typeset .wip > summary::before {
+background-color: rgb(255, 77, 6);
```
Comment (disrupted) [inline docs/docs/stylesheets/extra.css:38]: we could add [Malva](https://github.com/g-plane/malva) to our dprint config for formatting the CSS, wdyt? doesn't have to be part of this PR of course

**Review by disrupted — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml (hunk 15 above)
Comment (sujuka99) [reply .pre-commit-config.yaml:18]: They are excluded in `pyproject.toml`

File: docs/docs/stylesheets/extra.css (hunk 19 above)
Comment (sujuka99) [reply docs/docs/stylesheets/extra.css:38]: I support it

File: docs/docs/developer/contributing.md (hunk 18 above)
Comment (sujuka99) [reply docs/docs/developer/contributing.md:1]: Sounds better, will change

File: .github/workflows/release.yaml (hunk 14 above)
Comment (sujuka99) [reply .github/workflows/release.yaml:25]: indeed it should, fixed

File: docs/docs/developer/auto-generation.md (hunk 16 above)
Comment (sujuka99) [reply docs/docs/developer/auto-generation.md]: Nop, removed

**Review by disrupted — APPROVED** [review]
